bigtable/client/row_reader.cc

```diff
@@ -0,0 +1,188 @@
+
+// RowReader::iterator must satisfy the requirements of an InputIterator.
+namespace {
+using base_iterator = std::iterator<std::input_iterator_tag, Row>;
+
+static_assert(std::is_base_of<base_iterator, RowReader::iterator>::value,
+              "RowReader::iterator must be an InputIterator");
+static_assert(std::is_copy_constructible<RowReader::iterator>::value,
+              "RowReader::iterator must be CopyConstructible");
+static_assert(std::is_move_constructible<RowReader::iterator>::value,
+              "RowReader::iterator must be MoveConstructible");
+static_assert(std::is_copy_assignable<RowReader::iterator>::value,
+              "RowReader::iterator must be CopyAssignable");
+static_assert(std::is_move_assignable<RowReader::iterator>::value,
+              "RowReader::iterator must be MoveAssignable");
+static_assert(std::is_destructible<RowReader::iterator>::value,
+              "RowReader::iterator must be Destructible");
+
+static_assert(
+    std::is_convertible<decltype(*std::declval<RowReader::iterator>()),
+                        RowReader::iterator::value_type>::value,
+    "*it when it is of RowReader::iterator type must be convertible to "
+    "RowReader::iterator::value_type>");
+
+static_assert(std::is_same<decltype(++std::declval<RowReader::iterator>()),
+                           RowReader::iterator&>::value,
+              "++it when it is of RowReader::iterator type must be a "
+              "RowReader::iterator &>");
+
+}  // anonymous namespace
+
+RowReader::RowReader(
+    std::shared_ptr<DataClient> client, absl::string_view table_name,
+    RowSet row_set, std::int64_t rows_limit, Filter filter,
+    std::unique_ptr<RPCRetryPolicy> retry_policy,
+    std::unique_ptr<RPCBackoffPolicy> backoff_policy,
+    std::unique_ptr<internal::ReadRowsParserFactory> parser_factory)
+    : client_(std::move(client)),
+      table_name_(table_name),
+      row_set_(std::move(row_set)),
+      rows_limit_(rows_limit),
+      filter_(std::move(filter)),
+      retry_policy_(std::move(retry_policy)),
+      backoff_policy_(std::move(backoff_policy)),
+      context_(absl::make_unique<grpc::ClientContext>()),
+      parser_factory_(std::move(parser_factory)),
+      processed_chunks_count_(0),
+      rows_count_(0) {}
+
+RowReader::iterator RowReader::begin() {
+  if (not stream_) {
+    MakeRequest();
+  }
+  // Increment the iterator to read a row.
+  return ++internal::RowReaderIterator(this, false);
+}
+
+RowReader::iterator RowReader::end() {
+  return internal::RowReaderIterator(this, true);
+}
+
+void RowReader::MakeRequest() {
+  response_ = {};
+  processed_chunks_count_ = 0;
+
+  google::bigtable::v2::ReadRowsRequest request;
+  request.set_table_name(std::string(table_name_));
+
+  if (not last_read_row_key_.empty()) {
+    // There is a previous read row, so this is a restarted call
+    row_set_ = row_set_.Intersect(RowRange::Open(last_read_row_key_, ""));
```

Comment:
What happens if this returns an empty row set?  That means "everything" but should mean "we are done".

Reply:
That is an excellent question, and I've just committed a test that fails in such a scenario. Will follow up with the fix.

Reply:
In the interest of keeping other users from doing the same mistake, I've fixed this by handling the special case in RowSet. PTAL